File: 프로그래머스/3/214289. 에어컨/에어컨.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int temperature, int t1, int t2, int a, int b, vector<int> onboard) {
    const int OFFSET = 10;
    const int MIN_T = -10;
    const int MAX_T = 40;
    const int TSIZE = MAX_T - MIN_T + 1;
    const int INF = 1e9;
    
    int n = onboard.size();
    vector<vector<int>> dp(n, vector<int>(TSIZE, INF));
    dp[0][temperature - MIN_T] = 0;
    
    for (int i = 0; i < n - 1; i++) {
        for (int curT = MIN_T; curT <= MAX_T; curT++) {
            int curIdx = curT - MIN_T;
            int curCost = dp[i][curIdx];
            if (curCost == INF) continue;
            
            // 에어컨 끌 때
            int nextT = curT;
            if (curT < temperature) nextT++;
            else if (curT > temperature) nextT--;
            
            if (MIN_T <= nextT && nextT <= MAX_T) {
                if (onboard[i+1] == 0 || (t1 <= nextT && nextT <= t2)) {
                    int nextIdx = nextT - MIN_T;
                    dp[i+1][nextIdx] = min(dp[i+1][nextIdx], curCost);
                }
            }
            
            // 에어컨 킬 때
            for (int d = -1; d <= 1; d++) {
                int t2next = curT + d;
                if (t2next < MIN_T || t2next > MAX_T) continue;
                
                if (onboard[i+1] == 1 && !(t1 <= t2next && t2next <= t2)) continue;
                
                int addCost = (d == 0) ? b : a;
                int nextIdx = t2next - MIN_T;
                dp[i+1][nextIdx] = min(dp[i+1][nextIdx], curCost + addCost);
            }
        }
    }
    
    int answer = INF;
    for (int i = MIN_T; i <= MAX_T; i++) {
        answer = min(dp[n-1][i - MIN_T], answer);
    }
    return answer;
}

File: 프로그래머스/3/214289. 에어컨/에어컨.java
import java.util.*;
class Solution {
    static int INF = 1_000_000_000;
    static int OFFSET = 10;
    static int MAX = 51;
    public int solution(int temperature, int t1, int t2, int a, int b, int[] onboard) {
        int n = onboard.length;
        int[][] dp = new int[n][MAX];
        
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], INF);
        }
        
        dp[0][temperature + OFFSET] = 0;
        
        for (int time = 0; time < n - 1; time++) {
            for (int idx = 0; idx < MAX; idx++) {
                if (dp[time][idx] == INF) continue;
                int curTemp = idx - OFFSET;
                
                if (onboard[time] == 1 && (curTemp < t1 || curTemp > t2)) continue;
                
                // 1. 에어컨 OFF
                int nextTemp = curTemp;
                if (temperature < curTemp) nextTemp -= 1;
                else if (temperature > curTemp) nextTemp += 1;
                update(dp, time + 1, nextTemp + OFFSET, dp[time][idx]);
                
                // 2. 에어컨 ON - 온도 떨
                update(dp, time + 1, curTemp + OFFSET - 1, dp[time][curTemp + OFFSET] + a);
                
                // 3. 에어컨 ON - 온도 상
                update(dp, time + 1, curTemp + OFFSET + 1, dp[time][curTemp + OFFSET] + a);
                
                // 4. 에어컨 ON - 온도 유지
                update(dp, time + 1, curTemp + OFFSET, dp[time][curTemp + OFFSET] + b);
            }
        }
        
        int answer = INF;
        for (int idx = 0; idx < MAX; idx++) {
            int curTemp = idx - OFFSET;
            if (onboard[n - 1] == 1 && (curTemp < t1 || curTemp > t2)) continue;
            answer = Math.min(answer, dp[n - 1][idx]);
        }
        
        return answer;
    }
    
    private void update(int[][] dp, int time, int idx, int cost) {
        if (idx < 0 || idx >= MAX) return;
        dp[time][idx] = Math.min(dp[time][idx], cost);
    }
}